### PR TITLE
feat: add `Seq`-based iterators

### LINKED ISSUE #158: use `seq` for iterating over groups and matches
we may deprecate the `gen` API, or keep it.
I'll try to propose a PR when I have time.

### PR DESCRIPTION
**Pull request by c-cube**
Functional iterators are the future!

cc @Drup 
close #158

### CODE REVIEW

**Comment by Drup** [comment]
Would you advise to deprecate the `gen` versions ?

**Comment by c-cube** [comment]
I'm not sure, and it can be done in a separate commit.

**Comment by Drup** [comment]
@rgrinberg `seq` is the compatibility package that backports the `Seq` module to previous versions. It's a bit like the `result` or `bytes` packages. I believe it's an acceptable dependency.

**Comment by c-cube** [comment]
It's pretty weird @rgrinberg, there should be 2 versions of `seq`, one that depends on >= 4.07 and does nothing, and one that does something for older versions?

File: lib/core.ml [hunk 1]
```diff
@@ -1059,8 +1059,9 @@ module Mark = struct
 end
 
 type 'a gen = unit -> 'a option
+type 'a seq = 'a Seq.t
```
Comment (Drup) [inline lib/core.ml]: That type alias is really not useful here.

**Comment by Drup** [comment]
Let's keep the old ones for now!

**Pull request merged**